File: scala-stuff/src/main/scala/io/github/flyingcowmoomoo/arrowcatsfpguild/scala/EffectsSimple.scala
package io.github.flyingcowmoomoo.arrowcatsfpguild.scala

import cats.data.EitherT
import cats.effect.IO
import cats.effect.unsafe.implicits.global

object EffectsSimple {

  def main(args: Array[String]): Unit = { // The edge of our world
    val result = ioProgram().unsafeRunSync()
    println(result)
  }

  case class Id(id: Long)

  case class PersistenceError()

  case class User(email: String, name: String)

  case class ProcessedUser(id: Id, email: String, name: String)


  def fetchUser(): IO[Either[PersistenceError, User]] = {
    IO(Right(User("test@email.com", "Human")))
  }

  implicit class UserHelpers(val user: User) {
    def process(): IO[Either[PersistenceError, ProcessedUser]] = {
      if (user.email.matches("^(.+)@(.+)$")) {
        IO(Right(ProcessedUser(Id(1), user.email, user.name)))
      }
      else IO(Left(PersistenceError()))
    }
  }


  def ioProgram(): IO[Either[PersistenceError, ProcessedUser]] = {
    val result = for {
      user <- EitherT(fetchUser())
      processed <- EitherT(user.process())
    } yield processed
    result.value
  }
}

File: kotlin-stuff/src/main/kotlin/io/github/flyingcowmoomoo/kotlin/arrowcatsfpguild/EffectsSimple.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package io.github.flyingcowmoomoo.kotlin.arrowcatsfpguild

import arrow.core.Either
import arrow.core.Either.Left
import arrow.core.Either.Right
import arrow.core.computations.either
import io.github.flyingcowmoomoo.kotlin.arrowcatsfpguild.EffectsSimple.ioProgram
import kotlinx.coroutines.*

fun main(): Unit = runBlocking { // The edge of our world
    println(ioProgram())
}

object EffectsSimple {

    class Id(val id: Long)

    object PersistenceError

    data class User(val email: String, val name: String)

    data class ProcessedUser(val id: Id, val email: String, val name: String)

    suspend fun fetchUser(): Either<PersistenceError, User> =
        Right(User("test@email.com", "Human"))

    suspend fun User.process(): Either<PersistenceError, ProcessedUser> =
        when {
            email.contains(Regex("^(.+)@(.+)$")) -> Right(ProcessedUser(Id(1), email, name))
            else -> Left(PersistenceError)
        }

    suspend fun ioProgram(): Either<PersistenceError, ProcessedUser> =
        either {
            val user = fetchUser().bind()
            val processed = user.process().bind()
            processed
        }
}